### ISSUE TITLE
Python memory corruption crash

### ISSUE THREAD

**Issue by fmmoret** [shap/shap#1752]
I keep getting this error in the `_cext.dense_tree_shap` call when the tree explainer hits this line https://github.com/slundberg/shap/blob/54e6f0f10aa0442734aade259fc9712c4bf16537/shap/explainers/_tree.py#L389

I see a memory corruption error:
```
python(603,0x10c1c5dc0) malloc: Incorrect checksum for freed object 0x7fbd4d1dac00: probably modified after being freed.
Corrupt value: 0x7f2a00007f29
python(603,0x10c1c5dc0) malloc: *** set a breakpoint in malloc_error_break to debug
```
or a seg fault:
```
segmentation fault 
```

My dataframe that I give to the shap explainer can be as small as ~10 rows and still break.
I'm on a mac with python 3.7.8 and shap version 0.37.0 .

Would be happy to dig in further if I can get a guide for the cext dev setup. I haven't done c-ext work in python before, and I see that I can't just mess around with the installed lib in my python3.7/site-packages/shap folder -- only the built .so's are laying around after installation. 

I have been able to reproduce semi-consistently using my datasets, but not every time. I haven't been able to replicate with shap's datasets. 
On my side, my code looks like:

```
model = RandomForestRegressor(verbose=True, random_state=42, n_estimators=10)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)
fit = model.fit(X_train, y_train)
y_pred = model.predict(X_test)
explainer = shap.TreeExplainer(fit, data=X_test, check_additivity=False)
values = explainer.shap_values(X_test, approximate=False, check_additivity=False)
...
```

The issue happens for me in the "explainer.shap_values" line with a few of my datasets. It seems like it's more likely to crash when X has ~10+ columns instead of 1-3. However, it has no problems with `shap.datasets.adult()`

The X in my datasets is a pandas dataframe with column names that get ~50-70 chars long and have [a-zA-Z0-9()._] chars in any order and taken as user input.
@RAMitchell  -- I see you wrote most of the c-ext stuff. Would love tips for debugging.

**Comment by fmmoret** [comment]
This looks related: https://github.com/slundberg/shap/issues/954

**Comment by fmmoret** [comment]
Can confirm that setting max_depth=10 has fixed my problem too a la #954  -- albeit not ideal to constrain the model.
